kong/plugins/request-firewall/handler.lua

```diff
@@ -127,6 +153,85 @@ function plugin:access(config)
   if nil ~= path then
     path = string.gsub(path, "//", "/")
   end
+  
+  if config.graphql_match and config.graphql_match[path] then
+    local g_config = config.graphql_match[path]
+    local GqlParser = require('graphql-parser')
+    local parser = GqlParser:new()
+    local body, err, mimetype = kong.request.get_body()
+    local graph = parser:parse(body["query"])
+    
+    if graph:nestDepth() > g_config["nestDepth"] then
+      returnError(config, 'Nest depth exceeds limit') 
+      return
+    end
+    
+    --[[local path = ""
+    for a, b in pairs(graph) do
+      if nil ~= a then
+        path = path .. "a"..a .. " "
+      end
+      if nil ~= b.name then
+        path = path .. "b.name"..b.name .. " "
+      end
+      if nil ~= b.type then
+        path = path .. "b.type"..b.type .. " "
+      end
+    end
+    returnError(config, path)]]
+    
+    for _, op in pairs(graph) do
+      -- Check operation type e.g. query / mutation
+      
+      if nil ~= op.type then 
+      
+        if nil == g_config["structure"][op.type] then
+          returnError(config, 'Type '..op.type..' is not allowed')  
+          return
+        end
+        
+        for _,root in pairs(op.fields) do
+          -- Check RootElement name e.g. CreateToken
+          rootConfig = g_config["structure"][op.type][root.name]
+          if nil == rootConfig then
+            returnError(config, 'Root element '..root.name..' is not allowed')  
+            return
+          end
+          
+          -- Check variables by validateTable
+          local status, err = pcall(function() m.validateTable(config, rootConfig.variables, config.allow_unknown_body, root.name, body["variables"]) end)
+          if not status then
+            returnError(config, tostring(err.msg))
+            return
+          end
+          
+          for _, field in pairs(root.fields) do
+            -- Check field name e.g. token
+            if nil == rootConfig.subfields[field.name] then
+              returnError(config, 'Field '..root.name..'.'..field.name..' is not allowed')  
+              return
+            end
+          
+            -- Check sub-elements within the field
+            -- if utils.contains(rootConfig.subfields[field.name].enum, field.fields)
+            --if rootConfig.subfields[field.name] then
+            
+            
+            local status, err = pcall(function() checkSubElements(rootConfig.subfields[field.name].subElements, field.fields, field.name) end)
+            --checkSubElements(config, rootConfig.subfields[field.name].subElements, field.fields, field.name)
+            if not status then
+              returnError(config, tostring(err.msg))
+              return
+            end
+          end
+          
+        end
+      end
+    end
+    
+    returnError(config, '', 200)
```

Comment:
Should not return 200 in here. If you do, it will return 200 "without" sending the request to the downstream server.